File: java/src/main/java/topcoder/stockhistory/StockHistory.java
package topcoder.stockhistory;

/**
 * https://community.topcoder.com/stat?c=problem_statement&pm=3971&rd=6521
 */
public class StockHistory {
  public static int maximumEarnings(int initialInverstment, int monthlyContribution, String[] stockPrices) {
    int month = stockPrices.length;
    int corp = stockPrices[0].split(" ").length;

    int[][] prices = new int[month][corp];
    double max = 0;
    double profit = 0;
    double[] propertion = new double[month - 1];
    boolean[] buy = new boolean[month - 1];

    for (int i = 0; i < month; i++) {
      String[] s = stockPrices[i].split(" ");
      for (int j = 0; j < s.length; j++) {
        prices[i][j] = Integer.valueOf(s[j]);
      }
    }

    for (int i = month - 2; i >= 0; i--) {
      for (int j = 0; j < corp; j++) {
        double p = 1.0 * prices[month - 1][j] / prices[i][j] - 1.0;
        if(0 < p && max < p) {
          buy[i] = true;
          max = p;
          propertion[i] = p;
        }
      }
    }

    int money = initialInverstment;
    for(int i = 0; i < buy.length; i++ ) {
      if(buy[i]) {
        profit += money * propertion[i];
        money = 0;
      }
      money += monthlyContribution;
    }
    return (int)Math.round(profit);
  }
}

File: scala/src/main/scala/topcoder/stockhistory/StockHistory.scala
package topcoder.stockhistory

/**
  * https://community.topcoder.com/stat?c=problem_statement&pm=3971&rd=6521
  */
object StockHistory {
  def maximumEarnings(initialInverstment: Int, monthlyContribution: Int, stockPrices: Array[String]) = {
    var money = initialInverstment
    val stockPricesReverse = stockPrices.map(_.split(" ").map(_.toInt)).reverse
    val month = stockPricesReverse.length
    val corp = stockPricesReverse(0).length

    val lastMonthPrice = stockPricesReverse.head
    val proportion = stockPricesReverse.tail.foldLeft(0.0, List.empty[(Boolean, Double)]) {(acc, cur) => {
      cur.zip(lastMonthPrice).map{ case (c, l) => 1.0 * l/c - 1.0 }.max match {
        case x if x > acc._1 => (x, acc._2:+ (true, x))
        case x => (acc._1, acc._2:+ (false, x))
      }
    }}

    proportion._2.reverse.foldLeft(0.0){(acc, cur) =>
      if(cur._1) {
        val newAcc = acc + money * cur._2
        money = monthlyContribution
        newAcc
      } else {
        money += monthlyContribution
        acc
      }
    }.toInt
  }
}
